flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/ContinuousHiveSplitEnumerator.java

```diff
@@ -0,0 +1,230 @@
+			enumeratorContext.callAsync(
+					this::monitorAndGetSplits,
+					this::handleNewSplits,
+					discoveryInterval,
+					discoveryInterval);
+		} catch (Exception e) {
+			throw new FlinkHiveException("Failed to start continuous split enumerator", e);
+		}
+	}
+
+	@Override
+	public void handleSourceEvent(int subtaskId, SourceEvent sourceEvent) {
+		if (sourceEvent instanceof RequestSplitEvent) {
+			readersAwaitingSplit.put(subtaskId, ((RequestSplitEvent) sourceEvent).hostName());
+			assignSplits();
+		} else {
+			LOG.error("Received unrecognized event: {}", sourceEvent);
+		}
+	}
+
+	@Override
+	public void addSplitsBack(List<HiveSourceSplit> splits, int subtaskId) {
+		LOG.debug("Continuous Hive Source Enumerator adds splits back: {}", splits);
+		stateLock.writeLock().lock();
+		try {
+			splitAssigner.addSplits(new ArrayList<>(splits));
+		} finally {
+			stateLock.writeLock().unlock();
+		}
+	}
+
+	@Override
+	public void addReader(int subtaskId) {
+		// this source is purely lazy-pull-based, nothing to do upon registration
+	}
+
+	@Override
+	public PendingSplitsCheckpoint<HiveSourceSplit> snapshotState() throws Exception {
+		stateLock.readLock().lock();
+		try {
+			Collection<HiveSourceSplit> remainingSplits = (Collection<HiveSourceSplit>) (Collection<?>) splitAssigner.remainingSplits();
+			return new ContinuousHivePendingSplitsCheckpoint(remainingSplits, currentReadOffset, seenPartitions);
+		} finally {
+			stateLock.readLock().unlock();
+		}
+	}
+
+	@Override
+	public void close() throws IOException {
+		try {
+			fetcherContext.close();
+		} catch (Exception e) {
+			throw new IOException(e);
+		}
+	}
+
+	private Void monitorAndGetSplits() throws Exception {
+		stateLock.writeLock().lock();
+		try {
+			List<Tuple2<Partition, T>> partitions = fetcher.fetchPartitions(fetcherContext, currentReadOffset);
+			if (partitions.isEmpty()) {
+				return null;
+			}
+
+			partitions.sort(Comparator.comparing(o -> o.f1));
+			List<HiveSourceSplit> newSplits = new ArrayList<>();
+			// the max offset of new partitions
+			T maxOffset = currentReadOffset;
+			Set<List<String>> nextSeen = new HashSet<>();
+			for (Tuple2<Partition, T> tuple2 : partitions) {
+				Partition partition = tuple2.f0;
+				List<String> partSpec = partition.getValues();
+				if (seenPartitions.add(partSpec)) {
+					T offset = tuple2.f1;
+					if (offset.compareTo(currentReadOffset) > 0) {
+						nextSeen.add(partSpec);
+					}
+					if (offset.compareTo(maxOffset) > 0) {
+						maxOffset = offset;
+					}
+					LOG.info("Found new partition {} of table {}, generating splits for it",
+							partSpec, tablePath.getFullName());
+					newSplits.addAll(HiveSourceFileEnumerator.createInputSplits(
+							0, Collections.singletonList(fetcherContext.toHiveTablePartition(partition)), jobConf));
+				}
+			}
+			currentReadOffset = maxOffset;
+			splitAssigner.addSplits(new ArrayList<>(newSplits));
+			if (!nextSeen.isEmpty()) {
+				seenPartitions.clear();
```

Comment:
`seenPartitions` vs `distinctPartitions`
I prefer the latter, because it is only used to remove duplication, and does not save all the partitions we have seen

Reply:
The field is a Set, so I think calling it `distinct` is redundant. Besides, `seenPartitions` is better to indicate the purpose of the filed, which is to filter out partitions that we have processed. Perhaps I can rename it to `seenPartitionsForOffset`?

Reply:
Renamed to `seenPartitionsSinceOffset`. Hopefully that makes more sense.